### PR TITLE
Dual Optimizer for GAN class

### PR DESCRIPTION
**Pull request by hello-fri-end**
This PR is an attempt to finish the pending work in #1888. I've made the following TODOs list:

- [x] Dual Optimizer Train function for StandardGAN and DCGAN.
- [x] Update Tests for StandardGAN and DCGAN. 
- [ ] Dual Optimizer Train function for WGAN and WGAN-GP.
- [ ] Update Tests for WGAN and WGAN-GP.
- [ ] Remove boost visitor methods.
- [ ] Current implementation overrides shuffle parameter of the optimizer as false. Figure out a way to sync shuffling between discriminator and generator.

### CODE REVIEW

**Comment by zoq** [comment]
Great, thanks for looking into this.

**Comment by hello-fri-end** [comment]
@zoq, Here's what I tried for  implementing `Gradient Penalty`:

```
double epsilon = math::Random();
  predictors.cols(numFunctions, numFunctions + batchSize - 1) =
      (epsilon * currentInput) + ((1.0 - epsilon) * generatedData);
  responses.cols(numFunctions, numFunctions + batchSize - 1) =
      -arma::ones(1, batchSize);
  discriminator.Gradient(discriminator.parameter, numFunctions,
      interpGradientDiscriminator, batchSize);
  normGradientDiscriminator = arma::norm(interpGradientDiscriminator,2); 
  res += lambda * std::pow((normGradientDiscriminator - 1), 2);
```

But `interpGradientDiscrminator` is gradient with respect to parameters of the discriminator network while we want to calculate the gradient with respect to the mixed batch images. 

Equivalent pytorch Code:
```
        epsilon_shape = [real_data.shape[0]] + [1]*(real_data.dim() - 1)
        epsilon = torch.rand(epsilon_shape)
        epsilon = epsilon.to(fake_data.device, fake_data.dtype)
        real_data = real_data.to(fake_data.dtype)
        x_hat = epsilon * real_data + (1-epsilon) * fake_data.detach()
        x_hat.requires_grad = True
        logits = self.discriminator(x_hat, condition, landmarks)
        logits = logits.sum()
        grad = torch.autograd.grad(
            outputs=logits,
            inputs=x_hat,
            grad_outputs=torch.ones(logits.shape).to(fake_data.dtype).to(fake_data.device),
            create_graph=True
        )[0]
```
I'm not sure how to get the gradient with respect to the mixed batch.

And for the backward pass:

  // Backward pass of gradient penalty
```
  const double alpha = 1e-8;
  discriminator.error = (
                        ( 2/batchSize )
                        * lambda
                        * (normGradientDiscriminator -1)
                        * (interpGradientDiscriminator/
                             (normGradientDiscriminator + alpha) )
                        );

  discriminator.predictors = predictors.cols(numFunctions, numFunctions + batchSize -1);
  discriminator.Backward();
  discriminator.ResetGradients(gradientPenaltyDiscriminator);
  discriminator.Gradient(discriminator.Predictors().cols(numFunctions,
        numFunctions + batchSize -1));
  gradientDiscriminator += gradientPenaltyDiscriminator;
```

**Comment by hello-fri-end** [comment]
@zoq : Can you try to take a look at this please?

**Comment by zoq** [comment]
> But `interpGradientDiscrminator` is gradient with respect to parameters of the discriminator network while we want to calculate the gradient with respect to the mixed batch images.

Can you clarify what you mean with mixed batch, I'm not sure I get what you mean.

**Comment by hello-fri-end** [comment]
> > But `interpGradientDiscrminator` is gradient with respect to parameters of the discriminator network while we want to calculate the gradient with respect to the mixed batch images.
> 
> Can you clarify what you mean with mixed batch, I'm not sure I get what you mean.

I mean we need to compute gradient with respect to `x'` ,  where  `x' = ε * real_batch + ( 1 - ε ) * generated_batch` but calling the `Gradient()` method of the discriminator network gives us the gradient with respect to discriminator's parameters. Maybe the pytorch code I've posted in the comment above is more helpful.

**Comment by hello-fri-end** [comment]
keep open, please. 
@zoq Can you please take a look at the last comment i've left here?

**Comment by hello-fri-end** [comment]
@zoq: Can you take a look here, please?
